src/utils/utils.hpp

```diff
@@ -35,6 +38,33 @@ namespace parthenon {
 void ChangeRunDir(const char *pdir);
 void ShowConfig();
 
+// Lets us "printf" into a `std::string` safely. Much more readable
+// than, and just as safe as, streams. std::format (C++20) makes this
+// obselete.
+// Note this isn't very efficient, as it runs through the format
+// string 3 times:
+// 1. To compute string size
+// 2. To actually format the string into the buffer
+// 3. To copy the char* into a std::string
+template <typename... Args>
+std::string StringPrintf(const char *fmt, Args... args) {
+  // Get size of format string, including null terminator by passing
+  // in nullptr as target buffer
+  int size = std::snprintf(nullptr, 0, fmt, std::forward<Args>(args)...) + 1;
+  PARTHENON_REQUIRE(size > 0, "Valid format string");
+  // Can't safely write directly the pointer in a std::string, as
+  // internals of std::string not guranteed by standard.
+  std::vector<char> buffer(size);
+  char *pbuffer = buffer.data();
+  std::snprintf(pbuffer, size, fmt, std::forward<Args>(args)...);
+  // Return cast to string
+  return std::string(pbuffer);
+}
+template <typename... Args>
+std::string StringPrintf(const std::string &fmt, Args... args) {
+  return StringPrintf(fmt.c_str(), std::forward<Args>(args)...);
+}
```

Comment:
Would make it possible to write out a formatted error in one line instead of setting up the string stream first.

Reply:
I did start writing this macro, but it's a little bit more complicated than I thought it would be. The order of the arguments for the `impl::parthenon_fail` method needs to be swapped so that it can take a variatic pack as the final argument set. Not a huge change, but something to be aware of.